### PR TITLE
ECS framework (NovelRT::Ecs)

### PR DESCRIPTION
**Pull request by RubyNova**
This is the ECS framework for `NovelRT::Ecs`.

Changes are expected over time, and is blocked until #180 is merged.

### CODE REVIEW

File: include/NovelRT/Atom.h [hunk 1]
```diff
@@ -1,68 +1,97 @@
...
 
+#include <atomic>
+#include <cstddef>
+
 #ifndef NOVELRT_ATOM_H
 #define NOVELRT_ATOM_H
 
-#ifndef NOVELRT_H
-#error Please do not include this directly. Use the centralised header (NovelRT.h) instead!
-#endif
```
Comment (capnkenny) [inline include/NovelRT/Atom.h:8]: Needed to remove the centralised header warning here? I'm assuming this won't cause any issues down the line? :eyes:

File: include/NovelRT.h [hunk 2]
```diff
@@ -170,8 +170,6 @@ namespace NovelRT::SceneGraph {
  * Contains timing features, such as timers.
  */
 namespace NovelRT::Timing {
-  // Windows and some other platforms use 100ns ticks
-  static const uint64_t TicksPerSecond = 10'000'000;
```
Comment (capnkenny) [inline include/NovelRT.h:174]: Why was this removed?

File: include/NovelRT/Ecs/Catalogue.h [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        std::vector<EntityId> _createdEntitiesThisFrame;
+
+        public:
+        Catalogue(size_t poolId, ComponentCache& componentCache, EntityCache& entityCache) noexcept;
+
+        template<typename TComponent>
+        [[nodiscard]] auto GetComponentView()
+        {
+            return ComponentView<TComponent>(_poolId, _componentCache.GetComponentBuffer<TComponent>());
+        }
```
Comment (capnkenny) [inline include/NovelRT/Ecs/Catalogue.h:28]: I'm sorry if I keep forgetting - are we switching to Allman-style bracing for new files?

File: include/NovelRT/Ecs/ComponentBuffer.h [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        [[nodiscard]] auto begin() const noexcept
+        {
+            return _rootSet.cbegin();
+        }
+
+        [[nodiscard]] auto end() const noexcept
+        {
+            return _rootSet.cend();
+        }
+        // clang-format on
```
Comment (capnkenny) [inline include/NovelRT/Ecs/ComponentBuffer.h:91]: I'm assuming this is for future-proofing? (The `clang-format` mentions are what I'm specifically referring to here.)

File: scripts/build.sh [hunk 5]
```diff
@@ -139,9 +139,9 @@ function Test {
   pushd "$TestDir"
 
   if [ -z "$remaining" ]; then
-    ctest --build-config "$configuration" --output-on-failure
+    ctest --build-config "$configuration" --output-on-failure --VV
   else
-    ctest --build-config "$configuration" --output-on-failure "${remaining[@]}"
+    ctest --build-config "$configuration" --output-on-failure --VV "${remaining[@]}"
```
Comment (capnkenny) [inline scripts/build.sh:144]: Is this still required?

File: src/NovelRT/Ecs/Catalogue.cpp [hunk 6]
```diff
@@ -0,0 +1,29 @@
+// Copyright © Matt Jones and Contributors. Licensed under the MIT Licence (MIT). See LICENCE.md in the repository root for more information.
+
+#include <NovelRT/Ecs/Catalogue.h>
+
+namespace NovelRT::Ecs
+{
+    Catalogue::Catalogue(size_t poolId, ComponentCache& componentCache, EntityCache& entityCache) noexcept : _poolId(poolId), _componentCache(componentCache), _entityCache(entityCache), _createdEntitiesThisFrame(std::vector<EntityId>{})
+    {
+
```
Comment (capnkenny) [inline src/NovelRT/Ecs/Catalogue.cpp:9]: Nit: Extra space? 😄

File: src/NovelRT/Ecs/ComponentCache.cpp [hunk 7]
```diff
@@ -0,0 +1,29 @@
+// Copyright © Matt Jones and Contributors. Licensed under the MIT Licence (MIT). See LICENCE.md in the repository root for more information.
+
+#include <NovelRT/Ecs/ComponentCache.h>
+
+namespace NovelRT::Ecs
+{   
+    ComponentCache::ComponentCache(size_t poolSize) noexcept : _componentMap(std::unordered_map<ComponentTypeId, void*, AtomHashFunction>{}), _poolSize(poolSize), _bufferPrepEvent(Utilities::Event<const std::vector<EntityId>&>())
+    {
+
```
Comment (capnkenny) [inline src/NovelRT/Ecs/ComponentCache.cpp:9]: Nit: extra space? 😃

**Review by capnkenny — COMMENTED** [review]
Nicely done @RubyNova!
With that said.... had a couple questions for you to answer before I can grant thine approval 😛

File: include/NovelRT/Atom.h (hunk 1 above)
Comment (FiniteReality) [reply include/NovelRT/Atom.h:8]: I believe the endgame is to remove centralised header warnings from all the includes to allow individual includes, and provide the combined headers as convenience headers

File: include/NovelRT/Ecs/Catalogue.h [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+#include "ComponentCache.h"
+#include "ComponentView.h"
+#include "EcsUtils.h"
+#include "EntityCache.h"
+
+namespace NovelRT::Ecs
+{
+    class Catalogue
+    {
+        private:
```
Comment (FiniteReality) [inline include/NovelRT/Ecs/Catalogue.h:15]: IMO these labels should be at the same indentation level as the class, or the members indented one further.

File: src/NovelRT/Ecs/SystemScheduler.cpp [hunk 9]
```diff
@@ -0,0 +1,206 @@
+// Copyright © Matt Jones and Contributors. Licensed under the MIT Licence (MIT). See LICENCE.md in the repository root for more information.
+
+#include <NovelRT/Ecs/SystemScheduler.h>
+
+namespace NovelRT::Ecs
+{
+    SystemScheduler::SystemScheduler(uint32_t maximumThreadCount) noexcept : _entityCache(1),
```
Comment (FiniteReality) [inline src/NovelRT/Ecs/SystemScheduler.cpp:7]: Indentation here is different to other constructors; it should be made consistent.

File: include/NovelRT/Atom.h (hunk 1 above)
Comment (capnkenny) [reply include/NovelRT/Atom.h:8]: Ah okay - I'll resolve this then!

File: include/NovelRT.h (hunk 2 above)
Comment (capnkenny) [reply include/NovelRT.h:174]: Nvm I see it was relocated - disregard.

File: include/NovelRT/Ecs/Catalogue.h [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+            return std::make_tuple(ComponentView<TComponents>(_poolId, _componentCache.GetComponentBuffer<TComponents>())...);
+        }
+
+        [[nodiscard]] EntityId CreateEntity() noexcept;
+
+        void DeleteEntity(EntityId entity) noexcept;
+    };
+}
+
+#endif //!NOVELRT_ECS_CATALOGUE_H
```
Comment (FiniteReality) [inline include/NovelRT/Ecs/Catalogue.h:42]: Missing newline here? Some files have them, some don't

File: include/NovelRT/Atom.h [hunk 11]
```diff
@@ -1,68 +1,97 @@
...
 
+    operator uintptr_t() noexcept {
+      return _value;
+    }
+
     // TODO: These should be internal to NovelRT
 
-    static Atom getNextEventHandlerId() {
+    static Atom getNextEventHandlerId() noexcept {
       static std::atomic_uintptr_t _nextEventHandlerId(0);
```
Comment (FiniteReality) [inline include/NovelRT/Atom.h:54]: Not necessary for this, but thinking for the future:
```cpp
namespace NovelRT::Atoms {
    struct EventHandler { };
    struct FontSet { };
    // etc.
};
// ...
namespace NovelRT {
    template <typename AtomKind>
    static Atom GetAtom<AtomKind>()
    {
        // impl
    }
}
// e.g:
using namespace NovelRT;

auto atom = GetAtom<Atoms::EventHandler>();
auto atom2 = GetAtom<Atoms::FontSet>();
```

File: include/NovelRT/Ecs/Catalogue.h (hunk 3 above)
Comment (RubyNova) [reply include/NovelRT/Ecs/Catalogue.h:28]: Yes. :)

I did my best to write this feature under the new format, however until it is in I am bound to mess stuff up lol - I can't entirely remember every single convention so I'm hoping the new formatting stuff will just remember for me in the future

File: include/NovelRT/Ecs/ComponentBuffer.h (hunk 4 above)
Comment (RubyNova) [reply include/NovelRT/Ecs/ComponentBuffer.h:91]: You'd be assuming correctly. :D

File: scripts/build.sh (hunk 5 above)
Comment (RubyNova) [reply scripts/build.sh:144]: No idea - I don't even remember what it does. :^)

File: include/NovelRT/Ecs/Catalogue.h (hunk 8 above)
Comment (RubyNova) [reply include/NovelRT/Ecs/Catalogue.h:15]: I wasn't sure what was agreed upon for this in terms of the new coding conventions so I just sort of "left it" for now. Do you want me to change this still?

File: include/NovelRT/Ecs/Catalogue.h (hunk 10 above)
Comment (RubyNova) [reply include/NovelRT/Ecs/Catalogue.h:42]: This should be the end of the file. I've removed new lines after this in all the ECS files as of my latest commit if you want to check it again. :D

File: include/NovelRT/Ecs/Catalogue.h (hunk 8 above)
Comment (FiniteReality) [reply include/NovelRT/Ecs/Catalogue.h:15]: If the others are fine with it, I'll oblige - I'm not super sure what the new conventions are either if I'm honest :sweat_smile:

Comment (capnkenny) [reply include/NovelRT/Ecs/Catalogue.h:15]: I'm still working in that tbh @FiniteReality so I'll let this go as is - something I'll keep in mind for the clang-format / clang-tidy PRs :)

**Review by capnkenny — APPROVED** [review]
Take note of @FiniteReality's idea, but otherwise LGTM 🎉
